decisions/C-ManagerInterface.md

```diff
@@ -0,0 +1,205 @@
+# Initial C API Proposal
+
+This proposal is scoped to the C API bindings for the minimal 
+`ManagerInterface` methods that have been ported to C++ to date, 
+specifically `identifier()` and `displayName()`. Despite this tight
+restriction in scope there are several design decisions to be addressed.
+
+## Namespacing
+
+In C, exported symbols must be namespaced by prefixing the symbol name
+with some string.
+
+Since our top-level namespace includes a version tag, which is likely to
+change, we must use a macro to ease maintenance, e.g.
+
+```c++
+#define OPENASSETIO_NS(symbol) ...
+```
+such that `OPENASSETIO_NS(MyStruct)` expands to 
+`openassetio_v0_0_MyStruct`.
+
+## Strings
+
+Taking inspiration from [this conversation](https://github.com/TheFoundryVisionmongers/OpenAssetIO/pull/194#discussion_r794047468)
+we can bundle a string buffer and length in a single C struct for error
+messages and parameters, e.g.
+
+```c++
+typedef struct {
+  const size_t maxSize;
+  char* buffer;
+  size_t usedSize;
+} OPENASSETIO_NS(SimpleString);
+```
+`usedSize` is for optimisation and safety and can technically be removed
+if we're happy to assume null-termination and accept the performance
+implications.
+
+## Opaque handles
+
+The opaque handle pattern used commonly in C APIs involves an arbitrary
+pointer to some user data, often coded simply as a `void*`. To add some
+meagre type safety, we can define an opaque handle as a pointer to an
+incomplete and ultimately unused class. We then `reinterpret_cast` it 
+to/from our actual pointer type on the C++ side. E.g.
+
+```c++
+typedef struct OPENASSETIO_NS(managerAPI_ManagerInterface_t) *
+    OPENASSETIO_NS(managerAPI_ManagerInterface_h);
+```
+
```

Comment:
I imagine this boiler-plater pattern with an intermediate type would be further wrapped in a macro for ease of use? Might it be better to just use _p instead of _h? Handle has historically often meant **, whereas this is a *.

Reply:
Interesting point on `_p` vs `_h`, we'll re-evaluate that. `_h` has the advantage of matching the terminology of "opaque handle", maybe we should switch to "opaque pointer"?

I wondered about wrapping in a short-form macro, a couple of considerations though
* The short-form macro surely requires some level of namespacing regardless? I'm not sure what the appropriate short-form name should be, whilst still avoiding conflicts. I guess we could remove the namespace entirely for a short-form, then its up to the user to be grown up and ensure they don't mix headers that have conflicting symbol/macro names?
* We envision the C API being primarily a binding layer - we don't really expect many people to develop in pure C - so looking overly verbose is perhaps not a big problem, anyway.